New Datastore API and new Spine API support

dmdashenkov opened the pull request:
What's done:
- updated whole project to the new Google Cloud Datastore API;
- implemented new concepts from Spine (including, but not limited to `StandStorage`).

_Limitations._
`RecordStorage` and `StandStorage` support only `Integer`, `Long`, `String` and `Message` (aka protobuf type) types for identifiers. This means that custom defined as POJO or other special cases IDs can **not** be used.

dmdashenkov commented:
@dmytro-kashcheiev PTAL
Note: this may not be the last version of the PR (see my comment above).
Note: Travis isn't configured for this project right now (see issue #5).

dmytro-kashcheiev commented on build.gradle:
```diff
@@ -1,8 +1,10 @@
 project.ext {
-    SPINE_VERSION = '0.3-SNAPSHOT';
-    PROTOBUF_VERSION = '3.0.0-beta-2';
+    SPINE_VERSION = '0.6.2-SNAPSHOT';
```
the Spine-Snapshot versions not available in /spine repository. Please use /spine-snapshots, or switch to release version.

dmdashenkov replied:
Done

dmytro-kashcheiev commented on gcd/src/main/java/org/spine3/sample/gae/Application.java:
```diff
@@ -28,7 +28,7 @@
  * @author Alexander Litus
  */
-public class Application extends org.spine3.examples.aggregate.server.Application {
+public class Application /*extends org.spine3.examples.aggregate.server.Application*/ {
```
Please do not comment code with out comments or todo

dmdashenkov replied:
Here and further resolved by adding a TODO (we need `examples`repo updated to Spine v0.6.0 and published to Maven to run the code in `Application`).

dmytro-kashcheiev commented on gcd/src/main/java/org/spine3/sample/gae/Application.java:
```diff
@@ -42,7 +42,7 @@
     @SuppressWarnings("TypeMayBeWeakened") // not in this case
     private Application(DatastoreStorageFactory storageFactory) {
-        super(storageFactory);
+        //super(storageFactory);
```
Same as above

dmdashenkov replied:
Done

dmytro-kashcheiev commented on gcd/src/main/java/org/spine3/sample/gae/Application.java:
```diff
@@ -55,8 +55,8 @@ public static void main(String[] args) {
-        final Application app = new Application(storageFactory);
-        app.execute();
+        //final Application app = new Application(storageFactory);
+        //app.execute();
```
Same as above

dmdashenkov replied:
Done

dmytro-kashcheiev commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreProperties.java:
```diff
@@ -69,62 +65,37 @@ private DatastoreProperties() {
      * {@link org.spine3.protobuf.Timestamps#convertToDate(TimestampOrBuilder)}.
      */
-    /* package */ static Property.Builder makeTimestampProperty(TimestampOrBuilder timestamp) {
+    /* package */
```
Don't we put `/* package */` on the same line? If yes then check this class there are several such cases.

dmdashenkov replied:
Done

armiol replied:
Looks like it is "undone" again.

armiol replied:
> Looks like it is "undone" again.

Could you please put the `package` declaration to the same line as method declaration?

dmytro-kashcheiev commented on build.gradle:
```diff
@@ -117,10 +123,33 @@ subprojects {
+    "$pathPrefix/script/start-datastore$scriptExtension".execute()
+}
+
+
```
extra empty line?

dmytro-kashcheiev commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreStorageFactory.java:
```diff
@@ -42,51 +43,88 @@
+     *
+     * @param datastore   the {@link Datastore} implementation to use.
+     * @param multitenant shows if storage factory is configured to serve a multitenant application
+     * @return creates new factory instance.
```
By the current javadoc I will expect that this is void method.
Better something like : 
`@return new instance if the {@code  DatastoreStorageFactory}`

dmdashenkov replied:
Done

dmytro-kashcheiev commented on gcd/src/main/java/org/spine3/sample/gae/Application.java:
```diff
@@ -55,8 +55,8 @@ public static void main(String[] args) {
         final LocalDatastoreStorageFactory storageFactory = LocalDatastoreStorageFactory.getDefaultInstance();
         storageFactory.getOptions().setEventIteratorPageSize(ITERATOR_PAGE_SIZE);
```
Looks like, ITERATOR_PAGE_SIZE never used in the code. Please check and remove all code connected to this option if needed.

dmdashenkov replied:
As I see it, we provide `DatastoreStorageFactory.Options#getEventIteratorPageSize` as part of the API. I'll add a clarifying comment instead of removing.

dmytro-kashcheiev commented on gcd/src/main/java/org/spine3/server/storage/datastore/DsProjectionStorage.java:
```diff
@@ -35,21 +38,24 @@
 
```
Class level javadoc required.

dmdashenkov replied:
Done

dmytro-kashcheiev commented on gcd/src/main/java/org/spine3/server/storage/datastore/DsPropertyStorage.java:
```diff
@@ -57,25 +57,22 @@ private DsPropertyStorage(DatastoreWrapper datastore) {
-        final Any anyResult = entityToMessage(entity, ANY_TYPE_URL);
-        return Messages.fromAny(anyResult);
+        final Any anyResult = entityToMessage(response, ANY_TYPE_URL);
+        return AnyPacker.unpack(anyResult);
```
Have an explicit `result` before return.

dmytro-kashcheiev commented on gcd/src/main/java/org/spine3/server/storage/datastore/Entities.java:
```diff
@@ -0,0 +1,157 @@
+                .setValue(valueBytes)
+                .setTypeUrl(type.value())
+                .build();
+        return AnyPacker.unpack(wrapped);
```
Have an explicit `result` before return.

dmdashenkov commented:
@dmytro-kashcheiev 
I've addressed all your comments.
PTAL again.

dmytro-kashcheiev commented on gcd/src/test/java/org/spine3/sample/gae/ApplicationShould.java:
```diff
@@ -20,13 +20,15 @@
     private static final String[] ARGS = new String[0];
 
+    @Ignore
     @Test
```
I think it's not a good solution to use `@Ignore` with out comment why this test is ignored now.

dmytro-kashcheiev commented on gcd/src/main/java/org/spine3/server/storage/datastore/Entities.java:
```diff
@@ -0,0 +1,158 @@
+        final Any wrapped = AnyPacker.pack(message);
+        final byte[] messageBytes = wrapped.getValue().toByteArray();
+        final Blob valueBlob = Blob.copyFrom(messageBytes);
+        final Entity entity = Entity.builder(key)
```
Use chain-of-calls code conventions of spine.

dmytro-kashcheiev commented on gcd/src/main/java/org/spine3/server/storage/datastore/Entities.java:
```diff
@@ -0,0 +1,158 @@
+            final Blob value = entity.getBlob(VALUE_PROPERTY_NAME);
+            final ByteString valueBytes = ByteString.copyFrom(value.toByteArray());
+
+            final Any wrapped = Any.newBuilder()
```
Use chain-of-calls code conventions of spine.

dmytro-kashcheiev commented on gcd/src/main/java/org/spine3/server/storage/datastore/Entities.java:
```diff
@@ -0,0 +1,158 @@
+        final Blob value = entity.getBlob(VALUE_PROPERTY_NAME);
+        final ByteString valueBytes = ByteString.copyFrom(value.toByteArray());
+
+        final Any wrapped = Any.newBuilder()
```
Use chain-of-calls code conventions of spine.

dmytro-kashcheiev commented on gcd/src/main/java/org/spine3/server/storage/datastore/DsStandStorage.java:
```diff
@@ -0,0 +1,176 @@
+            public AggregateStateId apply(@Nullable Object input) {
+                checkNotNull(input, "String ID must not be null.");
+
+                return AggregateStateId.of(input, type);
```
Have explicit result before return.

dmytro-kashcheiev commented on gcd/src/main/java/org/spine3/server/storage/datastore/DsAggregateStorage.java:
```diff
@@ -89,34 +95,46 @@ public void writeEventCountAfterLastSnapshot(I id, int eventCount) {
-        final List<AggregateStorageRecord> records = entitiesToMessages(entityResults, TYPE_URL);
-        final Iterator<AggregateStorageRecord> recordsIterator = records.iterator();
-        return recordsIterator;
+        final Query<?> query = Query.entityQueryBuilder()
```
Use chain-of-calls code conventions of spine. In common there are lot of such situation. You can use format settings from spine core to find them all.

dmytro-kashcheiev commented on gcd/src/main/java/org/spine3/server/storage/datastore/DsCommandStorage.java:
```diff
@@ -20,41 +20,78 @@
 
-    /* package */ static CommandStorage newInstance(DatastoreWrapper datastore) {
-        return new DsCommandStorage(datastore);
+    /* package */
```
please put to same line.

dmytro-kashcheiev commented on gcd/src/main/java/org/spine3/server/storage/datastore/DsEventStorage.java:
```diff
@@ -20,100 +20,205 @@
+        }
+    };
+
+    /* package */
```
Please put to same line.

dmytro-kashcheiev commented on gcd/src/main/java/org/spine3/server/storage/datastore/DsPropertyStorage.java:
```diff
@@ -40,12 +40,12 @@
     private final DatastoreWrapper datastore;
 
-    /* package */ static DsPropertyStorage newInstance(DatastoreWrapper datastore) {
+    /* package */
```
please put to same line. You can use `/\* package \*/\n` such search regExp pattern to find all such situations.

dmdashenkov commented:
@dmytro-kashcheiev PTAL again
Changes since you've reviewed last time:
 - Fixed code style issues you've reported
 - Changed artifact version to `0.2`
Also note: @armiol suggested to rename Maven group ID to `org.spine3` and artifact to `gae-java` and to sync versioning with Spine core projects.

dmdashenkov commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreWrapper.java:
```diff
@@ -43,7 +43,7 @@
     // Default time to wait before each read operation to ensure the data is consistent.
     // NOTE: enabled only if {@link #shouldWaitForConsistency} is {@code true}.
-    private static final int CONSISTENCY_AWAIT_TIME_MS = 100;
+    private static final int CONSISTENCY_AWAIT_TIME_MS = 1000;
```
@armiol 
How can we be sure that this will work?
I thought we agreed to execute read operations transactionally to enforce strong consistency.

dmdashenkov commented:
@dmytro-kashcheiev , @armiol 
We finally come to the home scratch on our `gae-java` updates. 
Please, take your final look and let's proceed to merge.

dmdashenkov commented:
@dmytro-kashcheiev , @armiol PTAL

dmdashenkov commented:
@armiol Now all testing job on this repo should be done. Please, review the changes.

dmdashenkov commented:
@armiol PTAL

armiol commented on build.gradle:
```diff
@@ -45,27 +57,33 @@ subprojects {
 
     repositories {
-        mavenCentral()
+        //mavenCentral()
```
Required either commenting (e.g. why) or removal.

I also suspect this may have to be reverted back.

dmdashenkov replied:
Removed (moved to `allProjects` block).

armiol commented on build.gradle:
```diff
@@ -45,27 +57,33 @@ subprojects {
         // SLF4J
-        compile group: 'org.slf4j', name: 'slf4j-api', version: '1.7.12'
-        compile group: 'org.slf4j', name: 'slf4j-jdk14', version: '1.7.12'
+        compile group: 'org.slf4j', name: 'slf4j-api', version: '1.7.21'
```
As long as you modify it, could you please extract the repeated versions as constants?

dmdashenkov replied:
Done

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreStorageFactory.java:
```diff
@@ -38,55 +39,92 @@
+        return new DatastoreStorageFactory(datastore, multitenant);
+    }
+
+    @SuppressWarnings({"OverridableMethodCallDuringObjectConstruction", "OverriddenMethodCallDuringObjectConstruction"}) // Overriding used for testing
```
Please have a comment on suppressions.

dmdashenkov replied:
Done

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreStorageFactory.java:
```diff
@@ -38,55 +39,92 @@
+    private final boolean multitenant;
 
     /**
+     * Creates new instance of {@code DatastoreStorageFactory}.
```
Newline is needed before `<p>`.

dmdashenkov replied:
Done

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreStorageFactory.java:
```diff
@@ -95,28 +133,24 @@ public void close() throws Exception {
 
-            this.pageSize = pageSize;
-        }
+    private static class StandStorageRecord extends Entity<String, EntityStorageRecord> {
```
Please add class-level Javadocs. It is not totally clear why you would need such a class and why it is hidden from any user customisation.

dmdashenkov replied:
Done

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreStorageFactory.java:
```diff
@@ -38,55 +39,92 @@
+     * Creates new instance of {@code DatastoreStorageFactory}.
+     *
+     * @param datastore   the {@link Datastore} implementation to use.
+     * @param multitenant shows if storage factory is configured to serve a multitenant application
```
If we do not support multitenancy in the way GCD recommends, let's add a `todo` item for that.

As I recall, the multitenant GCD-based apps should use namespaces for `Entities` (and, thus, `Queries`). I don't think we do it.

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreWrapper.java:
```diff
@@ -20,220 +20,246 @@
-     * @param mutation the mutation to commit.
-     * @throws RuntimeException if {@link Datastore#commit(CommitRequest)} throws a {@link DatastoreException}
+     * @param entity new {@link Entity} to put into datastore
+     * @throws DatastoreException if an {@link Entity} with such {@link Key} already exists.
```
I suspect that duplication is not the only reason. What if the remote storage is unreachable (e.g. network problem)? Looks like we should get `DatastoreException` as well.

dmdashenkov replied:
I copied from the Datastore javadoc instead: 
`@throws DatastoreException upon failure.`

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreWrapper.java:
```diff
@@ -20,220 +20,246 @@
-    @SuppressWarnings("TypeMayBeWeakened") // no, it cannot
-    /* package */ void commit(Mutation.Builder mutation) {
-        commit(mutation.build());
+    @SuppressWarnings("WeakerAccess") // Can be used in future
```
Either `Part of API` — if it is really so, or make it private.

We do not rely on `future` in the design, as it is non-determined.

dmdashenkov replied:
Done

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreWrapper.java:
```diff
@@ -20,220 +20,246 @@
-    /* package */ Iterator<EntityResult> runQueryForIterator(Query query) {
-        return new PagingDatastoreIterator(query, this, datastoreOptions.getEventIteratorPageSize());
+    /**
+     * Retrieves an {@link Entity} with given key from datastore.
```
`the given key`.
Also either `Datastore` or `the datastore`. I'd prefer `Datastore` option.

dmdashenkov replied:
Done

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreWrapper.java:
```diff
@@ -20,220 +20,246 @@
+     * Retrieves an {@link Entity} with given key from datastore.
+     *
+     * @param key {@link Key} to search for.
+     * @return the {@link Entity} or {@code null} if there is no such a key.
```
`in case of no results for the key given`.

dmdashenkov replied:
Done

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreWrapper.java:
```diff
@@ -20,220 +20,246 @@
 
     /**
-     * Converts the given {@link Message} to the {@link Entity.Builder}.
+     * Deletes all existing {@link Entity Entities} with given keys.
```
`with the keys given`.

dmdashenkov replied:
Done

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreWrapper.java:
```diff
@@ -20,220 +20,246 @@
 
-        any.setTypeUrl(typeUrl);
+    /**
+     * Commits a transaction.
```
Here and in all similar cases.

The new line is required before `<p>`.

dmdashenkov replied:
Done

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DsCommandStorage.java:
```diff
@@ -20,41 +20,74 @@
     private final DatastoreWrapper datastore;
 
-    private final TypeName typeName;
+    private static final Function<Entity, CommandStorageRecord> RECORD_MAPPER = new Function<Entity, CommandStorageRecord>() {
```
Looks like not fitting into 120 symbols-per-row.

dmdashenkov replied:
Fixed

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DsEventStorage.java:
```diff
@@ -20,100 +20,293 @@
+        }
+    }
+
+    private static class EventFilterChecker implements Predicate<EventStorageRecord> {
```
Please have class-level Javadocs.

dmdashenkov replied:
Done

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DsEventStorage.java:
```diff
@@ -20,100 +20,293 @@
+            this.contextFieldFilters = eventFilter.getContextFieldFilterList();
+        }
+
+        private boolean isDefault() {
```
`EventFilterChecker#isDefault` seems weird. Either the method name is wrong, or the class serves a different purpose than declared by its name.

Even if we think of the class instance as of predicate, `Predicate#isDefault()`is still weird.

dmdashenkov replied:
By analogy with protobuf types (`Message::getDefaultInstance`) default is the state of the object when all fields are empty.

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DsRecordStorage.java:
```diff
@@ -0,0 +1,248 @@
+ * @author Dmytro Dashenkov
+ * @see DatastoreStorageFactory
+ */
+class DsRecordStorage<I> extends RecordStorage<I> {
```
`/* package */`?

dmdashenkov replied:
Done

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DsEventStorage.java:
```diff
@@ -20,100 +20,293 @@
  * @see DatastoreStorageFactory
- * @see LocalDatastoreStorageFactory
  */
 class DsEventStorage extends EventStorage {
```
`/* package */`?

dmdashenkov replied:
Done

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DsRecordStorage.java:
```diff
@@ -0,0 +1,248 @@
+    private static final String VERSION_KEY = "version";
+    private static final TypeUrl RECORD_TYPE_URL = TypeUrl.of(EntityStorageRecord.class);
+    private static final String KIND = RECORD_TYPE_URL.getSimpleName();
+    private static final String ID_CONVERSION_ERROR_MESSAGE
```
Please check 120 symbols-per-row in this class.

dmdashenkov replied:
Done

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DsRecordStorage.java:
```diff
@@ -0,0 +1,248 @@
+    private static final TypeUrl RECORD_TYPE_URL = TypeUrl.of(EntityStorageRecord.class);
+    private static final String KIND = RECORD_TYPE_URL.getSimpleName();
+    private static final String ID_CONVERSION_ERROR_MESSAGE
+            = "Entity had ID of an invalid type; could not parse ID from String. Note: custom convection is not supported. See Identifiers#idToString.";
```
What is `convection`?

armiol replied:
Also, consider making the message shorter.  Something like:

`Cannot convert String to Entity ID: its type is not supported. For supported types please see <package name here>Identifiers#...`.

dmdashenkov replied:
Fixed

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DsRecordStorage.java:
```diff
@@ -0,0 +1,248 @@
+        datastore.createOrUpdate(entity.build());
+    }
+
+    private static class IdRecordPair<I> {
```
Please have class-level Javadocs.

dmdashenkov replied:
Done

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DsStandStorage.java:
```diff
@@ -0,0 +1,182 @@
+import static com.google.common.base.Preconditions.checkState;
+
+/**
+ * GAE datastore implementation of {@link StandStorage}.
```
`Datastore`. 

This seem to be a common mistake. Please check the comments for it yourselves.

dmdashenkov replied:
Done

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DsStandStorage.java:
```diff
@@ -0,0 +1,182 @@
+
+                final Any wrappedState = input.getState();
+                final Message state = AnyPacker.unpack(wrappedState);
+
```
Please remove empty line.

dmdashenkov replied:
Done

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DsStandStorage.java:
```diff
@@ -0,0 +1,182 @@
+                final Message state = AnyPacker.unpack(wrappedState);
+
+                final TypeUrl recordType = TypeUrl.of(state.getDescriptorForType());
+
```
Please remove empty line.

dmdashenkov replied:
Done

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DsStandStorage.java:
```diff
@@ -0,0 +1,182 @@
+
+                final TypeUrl recordType = TypeUrl.of(state.getDescriptorForType());
+
+                return type.equals(recordType);
```
Have a `result` variable. BTW, it's also a common mistake in the changeset.

dmdashenkov replied:
Done

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/Entities.java:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    /**
+     * Retrieves a {@link List} of message of given type, assignable from {@code Message},
```
`messages`

dmdashenkov replied:
Done

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/Entities.java:
```diff
@@ -0,0 +1,158 @@
+    /**
+     * Retrieves a {@link List} of message of given type, assignable from {@code Message},
+     * from a collection of {@link Entity Entities}.
+     * <p>If passed {@link Entity} is {@code null}, a default instance for given type is returned.
```
`for the given`.

dmdashenkov replied:
Done

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/Entities.java:
```diff
@@ -0,0 +1,158 @@
+     * <p>If passed {@link Entity} is {@code null}, a default instance for given type is returned.
+     *
+     * @param entities a collection of the source {@link Entity Entities} to get message form.
+     * @param type     {@link TypeUrl} of required message.
```
As I recall, we don't put dots at the end of `@param` description.

dmdashenkov replied:
Fixed

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/Entities.java:
```diff
@@ -0,0 +1,158 @@
+    /*package*/ static <M extends Message> M defaultMessage(TypeUrl type) {
+        final Class<M> messageClass = Messages.toMessageClass(type);
+        checkState(messageClass != null, String.format(
+                "Not found class for type url \"%s\". Try to rebuild project",
```
`the project`.

dmdashenkov replied:
Done

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/IdTransformer.java:
```diff
@@ -0,0 +1,197 @@
+ * @author Dmytro Dashenkov
+ */
+@SuppressWarnings("UtilityClass")
+/*package*/ class IdTransformer {
```
Class-level description is required.

dmdashenkov replied:
Done

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/IdTransformer.java:
```diff
@@ -0,0 +1,197 @@
+    }
+
+
+    private static boolean isNumber(Class<?> clss) {
```
The name is wrong. `isNumber` means `capture<? extends Number>`. 
I'd suggest `isOfSupportedNumberType`.

Instead of `clss` please use `clazz`, which is a common replacement for a keyword.

Also the implementation is wrong. `double` is a primitive. But I don't think that's what you want to return `true`.

dmdashenkov replied:
Fixed. Now the method checks only `Integer.class`, `Long.class`, `Integer#TYPE` and `Long#TYPE`.

armiol commented on gcd/src/main/resources/spine-dev-62685282c0b9.json:
```diff
@@ -0,0 +1,12 @@
+{
```
Why is it in the `main` resource folder, not in the `test` one?

dmdashenkov replied:
Moved

armiol commented on gcd/src/test/java/org/spine3/server/storage/datastore/EntitiesUtilShould.java:
```diff
@@ -0,0 +1,80 @@
+ * @author Dmytro Dashenkov
+ */
+@SuppressWarnings("InstanceMethodNamingConvention")
+public class EntitiesUtilShould {
```
You are testing `Entities`, but still call the class `EntitiesUtilShould`.

Please fix here and check whether you have more cases like this.

dmdashenkov replied:
Renamed

armiol commented on gcd/src/test/java/org/spine3/server/storage/datastore/TestDatastoreWrapper.java:
```diff
@@ -0,0 +1,189 @@
+import java.util.List;
+
+/**
+ * Custom extension of the {@link DatastoreWrapper} aimed on testing purposes.
```
`aimed on testing purposes` -> `for the integration testing.`

dmdashenkov replied:
Fixed

dmdashenkov commented:
@armiol PTAL again

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreStorageFactory.java:
```diff
@@ -38,55 +39,98 @@
+    /**
+     * Creates new instance of {@code DatastoreStorageFactory}.
+     *
+     * @param datastore   the {@link Datastore} implementation to use.
```
Again, I don't think we put dots at the end of parameter docs.

Please check here and through the codebase.

dmdashenkov replied:
Done now.

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreStorageFactory.java:
```diff
@@ -38,55 +39,98 @@
+
+    // Overriding used for testing
+    @SuppressWarnings({"OverridableMethodCallDuringObjectConstruction", "OverriddenMethodCallDuringObjectConstruction"})
+    /* package */DatastoreStorageFactory(Datastore datastore, boolean multitenant) {
```
Space is missing after the `/* package */` declaration.

dmdashenkov replied:
Fixed.

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreWrapper.java:
```diff
@@ -20,220 +20,252 @@
-    /* package */ Iterator<EntityResult> runQueryForIterator(Query query) {
-        return new PagingDatastoreIterator(query, this, datastoreOptions.getEventIteratorPageSize());
+    /**
+     * Retrieves an {@link Entity} with given key from the Datastore.
```
`the given key`.

dmdashenkov replied:
Fixed.

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreWrapper.java:
```diff
@@ -20,220 +20,252 @@
 
     /**
-     * Converts the given {@link EntityResultOrBuilder} to the {@link Message}.
+     * Deletes all existing {@link Entities} of given kind.
```
`of a kind given`.

dmdashenkov replied:
Fixed

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreWrapper.java:
```diff
@@ -20,220 +20,252 @@
-     * @param typeUrl the type url of the message
-     * @return the deserialized message
-     * @see Any#getTypeUrl()
+     * @param table Kind (a.k.a. type, table, etc.) of the records to delete.
```
Looks like it should be `kind` instead of `Kind`.

dmdashenkov replied:
Fixed

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreWrapper.java:
```diff
@@ -20,220 +20,252 @@
+    /**
+     * Starts a transaction.
+     *
+     * <p>Since this method is called and until one of {@link #commitTransaction()} or {@link #rollbackTransaction()}
```
`<p>After this method is called, all {@code Entity} modifications performed through this instance of {@code DatastoreWrapper}  become transactional. This behaviour lasts until either {@link #commitTransaction()} or {@link #rollbackTransaction()} is called.`

dmdashenkov replied:
Fixed.

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreWrapper.java:
```diff
@@ -20,220 +20,252 @@
+    /**
+     * Commits a transaction.
+     *
+     * <p>All transactional operations are being performed.
```
`<p>Upon the method call, all the modifications within the active transaction are applied.`

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreWrapper.java:
```diff
@@ -20,220 +20,252 @@
+     *
+     * <p>All transactional operations are being performed.
+     *
+     * <p>All next operations become non-transactional until {@link #startTransaction()} is called again
```
`<p>After this method execution is over, all the further modifications made through the current instance of {@code DatastoreWrapper} become non-transactional.`

And let's avoid `is called again`. It is clear enough.

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreWrapper.java:
```diff
@@ -20,220 +20,252 @@
+     *
+     * <p>All next operations become non-transactional until {@link #startTransaction()} is called again
+     *
+     * @throws IllegalStateException if no transaction is started on this instance of {@code DatastoreWrapper}.
```
`if no transaction is active for the current instance of {@code DatastoreWrapper}`.

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreWrapper.java:
```diff
@@ -20,220 +20,252 @@
 
-        any.setTypeUrl(typeUrl);
+    /**
+     * Rollback a transaction.
```
`Rollbacks`.

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreWrapper.java:
```diff
@@ -20,220 +20,252 @@
+    /**
+     * Rollback a transaction.
+     *
+     * <p>All transactional operations are not performed.
```
<p>The invocation of this method does NOT apply any of modifications made in scope of the active transaction. Instead, they become lost forever.`

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreWrapper.java:
```diff
@@ -20,220 +20,252 @@
+     *
+     * <p>All transactional operations are not performed.
+     *
+     * <p>All next operations become non-transactional until {@link #startTransaction()} is called again
```
`<p>After this method execution is over, all the further modifications made through the current instance of {@code DatastoreWrapper} become non-transactional.`

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreWrapper.java:
```diff
@@ -20,220 +20,252 @@
+     *
+     * <p>All next operations become non-transactional until {@link #startTransaction()} is called again
+     *
+     * @throws IllegalStateException if no transaction is started on this instance of {@code DatastoreWrapper}.
```
`if no transaction is active for the current instance of {@code DatastoreWrapper}.`

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreWrapper.java:
```diff
@@ -20,220 +20,252 @@
+    /*package*/ void commitTransaction() throws IllegalStateException {
+        checkState(isTransactionActive(), ACTIVE_TRANSACTION_CONDITION_MESSAGE);
+        activeTransaction.commit();
+        actor = datastore;
```
To emphasise the change of the `actor` value, let's write `this.actor = datastore`.
Please apply here and in related cases.

Note that calls of `actor` methods are fine without `this`.

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DsEventStorage.java:
```diff
@@ -20,100 +20,295 @@
+        final PropertyFilter greaterThen = PropertyFilter.gt(TIMESTAMP_NANOS_PROPERTY_NAME, lower);
+        final PropertyFilter lessThen = PropertyFilter.lt(TIMESTAMP_NANOS_PROPERTY_NAME, upper);
+        final CompositeFilter filter = CompositeFilter.and(greaterThen, lessThen);
+        return Query.newEntityQueryBuilder()
```
Please have a `result` value before return.

Usage of this pattern through the code will make the debugging much easier.

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/DsStandStorage.java:
```diff
@@ -0,0 +1,181 @@
+import static com.google.common.base.Preconditions.checkState;
+
+/**
+ * GAE Datastore implementation of {@link StandStorage}.
```
In other implementations we say `Google Cloud Datastore`. Let's call it in this way here as well.

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/Entities.java:
```diff
@@ -0,0 +1,158 @@
+
+    /**
+     * Retrieves a message of given type, assignable from {@code Message}, from an {@link Entity}.
+     * <p>If passed {@link Entity} is {@code null}, a default instance for the given type is returned.
```
Newline is missing before `<p>`.

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/Entities.java:
```diff
@@ -0,0 +1,158 @@
+    /**
+     * Retrieves a {@link List} of messages of given type, assignable from {@code Message},
+     * from a collection of {@link Entity Entities}.
+     * <p>If passed {@link Entity} is {@code null}, a default instance for given type is returned.
```
Newline is missing before `<p>`.

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/IdTransformer.java:
```diff
@@ -0,0 +1,232 @@
+    }
+
+    /**
+     * Transforms given wildcard ID into it's string representation.
```
`its`.
`String`?

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/IdTransformer.java:
```diff
@@ -0,0 +1,232 @@
+     *           </ul>
+     *
+     *           <p>Note: there are some limitations on what a string ID can be.
+     *           See <a href="https://cloud.google.com/appengine/docs/python/datastore/entities>Datastore docs</a>
```
Why reference a Python docs? I am pretty sure there is Java version.

dmdashenkov replied:
This does not actually reference to Python docs but to common docs containing basic Datastore concepts. We do have Java docs, but they are not equivalent but may contain similar but not the same info split into articles in some other way.
That article was the best match for this context so I used it.

armiol commented on gcd/src/main/java/org/spine3/server/storage/datastore/IdTransformer.java:
```diff
@@ -0,0 +1,232 @@
+     * Transforms ID from a {@code String} representation back to a {@link Number} or a {@link Message protobuf message}.
+     *
+     * @param stringId          the {@code String} representation of the ID
+     * @param parametrizedClass {@link Class parametrized type} of the {@link org.spine3.server.entity.Entity} to restore ID for
```
Please check 120 symbols-per-row.

armiol commented on spine-dev-62685282c0b9.json:
```diff
@@ -0,0 +1,12 @@
+{
```
As of second thought, this file must NOT be present in the codebase.

Instead, it should be stored encrypted. And decrypted before the unit test start.

armiol replied:
Use [this guide](https://docs.travis-ci.com/user/encrypting-files/).

dmdashenkov commented:
@armiol **PTAL** again.
Also __note:__ sometimes the tests may fail due to our mechanism of enforcing the strong consistency. If this happens, a Travis build should be restarted. This increases the chance for the data to be consistent (with each next check).

dmdashenkov commented:
@dmytro-kashcheiev PTAL
